fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java

```diff
@@ -363,20 +367,50 @@ private static void timerIncrement(@Nullable FDBStoreTimer timer, FDBStoreTimer.
         // Need to do this each transaction because other index enabled state might have changed. Could cache based on that.
         // Copying the state also guards against changes made by other online building from check version.
         // TODO: need some state to avoid generating the same synthetic record via more than one self-join path for non-idempotent indexes.
+        AtomicLong recordsScannedCounter = new AtomicLong();
+
+        final AtomicReference<RecordCursorResult<T>> nextResult = new AtomicReference<>(null);
         return AsyncUtil.whileTrue(() -> cursor.onNext().thenCompose(result -> {
-            if (!result.hasNext()) {
+            RecordCursorResult<T> currResult;
+            boolean isExhausted = false;
+            if (result.hasNext()) {
+                // has next, process one previous item (if exists)
+                currResult = nextResult.get();
+                nextResult.set(result);
+                if (currResult == null) {
+                    // that was the first item, nothing to process
+                    return AsyncUtil.READY_TRUE;
+                }
+            } else {
                 // end of the cursor list
                 timerIncrement(timer, FDBStoreTimer.Counts.ONLINE_INDEX_BUILDER_RANGES_BY_COUNT);
-                lastResultSet.accept(result);
-                return AsyncUtil.READY_FALSE;
+                if (result.getNoNextReason() != SOURCE_EXHAUSTED) {
+                    // should we trigger a warning (or an exception) if nextResult is null - i.e. non-exhausted empty list?
```

Comment:
Oh, wait, isn't this hit in the current code, too, though? In particular, whenever we hit an _in-band_ limit, e.g., the "returned record limit". So this code is actually probably fine as is. I guess the only thing would be if wanted to pro-actively warn us if we hit an out-of-band limit. Given how the cases (in-band and out-of-band) are actually set up so that they should be handled the same, I'm tempted to say we don't warn and just leave this as is (maybe without this comment).

Reply:
Removed the comment...